### PR TITLE
core[minor]: Add msg content formatting util

### PR DESCRIPTION
**Pull request by baskaryan**
Fixes #25813 

Direct usage
```python
from langchain_core.messages import format_messages_as, AIMessage, HumanMessage, ToolMessage, SystemMessage

messages = [
    SystemMessage([{"type": "text", "text": "foo"}]),
    {"role": "user", "content": [{"type": "text", "text": "whats in this"}, {"type": "image_url", "image_url": {"url": "data:image/png;base64,'/9j/4AAQSk'"}}]},
    AIMessage("", tool_calls=[{"name": "analyze", "args": {"baz": "buz"}, "id": "1", "type": "tool_call"}]),
    ToolMessage("foobar", tool_call_id="1", name="bar"), 
    {"role": "assistant", "content": "thats nice"},
]
oai_strings = format_messages_as(messages, format="openai", text="string")
# -> [
#     SystemMessage(content='foo'),
#     HumanMessage(content=[{'type': 'text', 'text': 'whats in this'}, {'type': 'image_url', 'image_url': {'url': "data:image/png;base64,'/9j/4AAQSk'"}}]),
#     AIMessage(content='', tool_calls=[{'name': 'analyze', 'args': {'baz': 'buz'}, 'id': '1', 'type': 'tool_call'}]),
#     ToolMessage(content='foobar', name='bar', tool_call_id='1'),
#     AIMessage(content='thats nice')
# ]

anthropic_blocks = format_messages_as(messages, format="anthropic", text="block")
# -> [
#     SystemMessage(content=[{'type': 'text', 'text': 'foo'}]),
#     HumanMessage(content=[{'type': 'text', 'text': 'whats in this'}, {'type': 'image', 'source': {'type': 'base64', 'media_type': 'image/png', 'data': "'/9j/4AAQSk'"}}]),
#     AIMessage(content=[{'type': 'tool_use', 'input': {'baz': 'buz'}, 'id': '1', 'name': 'analyze'}], tool_calls=[{'name': 'analyze', 'args': {'baz': 'buz'}, 'id': '1', 'type': 'tool_call'}]),
#     HumanMessage(content=[{'type': 'tool_result', 'content': 'foobar', 'tool_use_id': '1', 'is_error': False}]),
#     AIMessage(content=[{'type': 'text', 'text': 'thats nice'}])
# ]
```

Chaining
```python
  from langchain_core.messages import format_messages
  from langchain.chat_models import init_chat_model
  formatter = format_messages(format="langchain-openai", text_format="string")
  llm = init_chat_model() | formatter
  llm.invoke(
      [{"role": "user", "content": "how are you"}],
      config={"model": "gpt-4o"},
  )
  # -> AIMessage(["I am good..."], ...)
  llm.invoke(
      [{"role": "user", "content": "whats your name"}],
      config={"model": "claude-3-5-sonnet-20240620"}
  )
  # -> AIMessage(["My name is...], ...)
```

### CODE REVIEW

File: libs/core/langchain_core/messages/utils.py [hunk 1]
```diff
@@ -845,6 +909,610 @@ def list_token_counter(messages: Sequence[BaseMessage]) -> int:
...
+) -> Union[BaseMessage, List[BaseMessage]]:
+    """Convert message contents into a standard format.
+
+    .. versionadded:: 0.2.36
+
+    Args:
+        messages: Message-like object or iterable of objects whose contents are already
+            in OpenAI, Anthropic, Bedrock Converse, or VertexAI formats.
+        format: Format to convert message contents to.
+        text: How to format text contents. If ``text='string'``  then any string
```
Comment (vbarda) [inline libs/core/langchain_core/messages/utils.py:927]: could we do `content_format` or something instead of `text`?

Comment (vbarda) [reply libs/core/langchain_core/messages/utils.py:927]: and the above could be "provider" or something

Comment (baskaryan) [reply libs/core/langchain_core/messages/utils.py:927]: content_format feels a little too broad given it only relates to text blocks — text_format?

**Comment by mschoenb97IL** [comment]
Looks excellent, thank you!

**Comment by mschoenb97IL** [comment]
Looks good! But does this support message chunk types? (i.e. AIMessageChunk, ToolMessageChunk)

**Comment by baskaryan** [comment]
> Looks good! But does this support message chunk types? (i.e. AIMessageChunk, ToolMessageChunk)

as currently implemented it can format a chunk one at a time but it can't transform a stream on the fly. meaning if you used in a chain and you tried to stream from that chain, you would just get a single accumulated chunk out instead of token-level chunks

will think about if/how we can support transforming a stream

**Comment by mschoenb97IL** [comment]
> > Looks good! But does this support message chunk types? (i.e. AIMessageChunk, ToolMessageChunk)
> 
> as currently implemented it can format a chunk one at a time but it can't transform a stream on the fly. meaning if you used in a chain and you tried to stream from that chain, you would just get a single accumulated chunk out instead of token-level chunks
> 
> will think about if/how we can support transforming a stream

So if I have a downstream application that consumes message chunks from a graph, can I apply this function to uniformize chunks across different providers?

File: libs/core/langchain_core/messages/utils.py (hunk 1 above)
Comment (vbarda) [reply libs/core/langchain_core/messages/utils.py:927]: sure!

**Comment by jacoblee93** [comment]
What if you're going from more expressive to less expressive (e.g. Anthropic blocks -> OpenAI strings)?

If we are dropping expressiveness in that case, how about `coerce_messages_to` intead?

File: libs/core/langchain_core/messages/utils.py [hunk 2]
```diff
@@ -845,6 +862,775 @@ def list_token_counter(messages: Sequence[BaseMessage]) -> int:
...
+                    yield msg
+
+            return RunnableGenerator(partial(transform, **kwargs), name=func.__name__)
+
+    wrapped.__doc__ = func.__doc__
+    return wrapped
+
+
+@_runnable_generator
+def format_messages_as(
```
Comment (efriis) [inline libs/core/langchain_core/messages/utils.py:937]:
```suggestion
def format_messages(
```
? 🌝

File: libs/core/langchain_core/messages/utils.py [hunk 3]
```diff
@@ -845,6 +862,775 @@ def list_token_counter(messages: Sequence[BaseMessage]) -> int:
...
+    messages: Union[MessageLikeRepresentation, Sequence[MessageLikeRepresentation]],
+    *,
+    format: Literal["openai", "anthropic"],
+    text_format: Literal["string", "block"],
```
Comment (efriis) [inline libs/core/langchain_core/messages/utils.py:941]:
```suggestion
    format: Literal["openai", "anthropic", "openai_text", "anthropic_text"],
```
might be confusing to have `text_format` option if we ever introduce something else that doesn't offer one or the other

**Comment by baskaryan** [comment]
> What if you're going from more expressive to less expressive (e.g. Anthropic blocks -> OpenAI strings)?
> 
> If we are dropping expressiveness in that case, how about `coerce_messages_to` intead?

only converts to strings if its all text blocks

File: libs/core/langchain_core/messages/utils.py (hunk 2 above)
Comment (efriis) [reply libs/core/langchain_core/messages/utils.py:937]: don't we also have a `convert_messages_to_dict` somewhere that does the same thing? Would that kind of naming make more sense?

Might make it clearer this is only meant to handle conversion to dicts (incase anyone tries to add conversions to custom sdks at some point)

**Review by efriis — APPROVED** [review]
light preferences commented. Also looks good as is

File: libs/core/langchain_core/messages/utils.py (hunk 3 above)
Comment (baskaryan) [reply libs/core/langchain_core/messages/utils.py:941]: something that doesn't offer string or block options?

File: libs/core/langchain_core/messages/utils.py (hunk 2 above)
Comment (baskaryan) [reply libs/core/langchain_core/messages/utils.py:937]: watcha mean meant to handle conversion to dict?

File: libs/core/langchain_core/messages/utils.py (hunk 3 above)
Comment (efriis) [reply libs/core/langchain_core/messages/utils.py:941]: yep - for example if new_model_provider_c only supports blocks and not string content in their api for some reason. Or anthropic introduces a block_2 format. Keeping it 1-dimensional is easier to maintain in that case.

File: libs/core/langchain_core/messages/utils.py (hunk 2 above)
Comment (efriis) [reply libs/core/langchain_core/messages/utils.py:937]: it formats the contents as openai/anthropic dicts, not <insert some model provider sdk> basemodel/protobuf/etc types for theirs.

**Comment by mschoenb97IL** [comment]
@baskaryan Are there pending changes here? Would love to be able to use this soon. Thanks again!

**Comment by mschoenb97IL** [comment]
@baskaryan Why was this closed?

**Pull request closed without merging**